io_uring: implement automatic buffer selection

vrischmann opened the pull request:
This implements [automatic buffer selection](https://lwn.net/Articles/815491/) for `io_uring`.

The current api of `read` and `recv` is not great when using automatic buffer selection, we have to write relatively ugly code like this:
```zig
var slice: []u8 = "";
slice.len = 128;

var sqe = try ring.read(0xdfdfdfdf, fd, slice[0..], 0);
sqe.flags |= linux.IOSQE_BUFFER_SELECT;
sqe.buf_index = group_id;
```
because `read` takes a slice and gives its `ptr` and `len` to the kernel via [these fields](https://github.com/ziglang/zig/blob/master/lib/std/os/linux.zig#L3503-L3504).

This is the reason why I added `read_raw` and `recv_raw` which take a `?[*]u8` and a separate length, so we can write code like this instead:
```zig
var sqe = try ring.read_raw(0xdfdfdfdf, fd, null, buffer_len, 0);
sqe.flags |= linux.IOSQE_BUFFER_SELECT;
sqe.buf_index = group_id;
```
which is better imo.

I'm not 100% happy with this API either, because:
* it feels redundant given the super small code difference between the two methods
* future operations which take a buffer as input and support automatic buffer selection will also need to implement two different methods.

There are other solutions but they require breaking existing code using `read` and `recv`.

1)  define `read` the same as `read_raw`
```zig
pub fn read(self: *IO_Uring, user_data: u64, fd: os.fd_t, buffer: ?[*]u8, len: usize, offset: u64) !*io_uring_sqe
```
and let the caller give the correct pointer/len.

2) we could add a buffer type, something like this:
```zig
pub const ReadBuffer = union(enum) {
    slice: []u8,
    buffer_select: usize,
};

pub fn read(self: *IO_Uring, user_data: u64, fd: os.fd_t, buffer: ReadBuffer, offset: u64) !*io_uring_sqe

var buffer: [128]u8 = undefined;
var sqe = try ring.read(0, fd, .{ .slice = &buffer }, 0);

var sqe = try ring.read(0, fd, .{ .buffer_select = buffer_len }, 0);
```

Any input on this is appreciated, I might have missed something obvious.

andrewrk commented:
This old PR had an Azure CI failure and the log was deleted since it is so old, so I rebased this against master.

jorangreef commented on lib/std/os/linux/io_uring.zig:
```diff
@@ -857,6 +895,41 @@ pub const IO_Uring = struct {
+        self: *IO_Uring,
+        user_data: u64,
+        buffers: [*]u8,
+        num: usize,
```
```suggestion
        buffers_count: usize,
```

jorangreef commented on lib/std/os/linux/io_uring.zig:
```diff
@@ -1508,6 +1592,28 @@ pub fn io_uring_prep_linkat(
     sqe.rw_flags = flags;
 }
 
+pub fn io_uring_prep_provide_buffers(
```
Nice to see the same prep method names coming through from liburing.

jorangreef commented on lib/std/os/linux/io_uring.zig:
```diff
@@ -522,6 +541,25 @@ pub const IO_Uring = struct {
+    ///
+    /// The only difference with `recv` is that this takes a nullable many-item pointer and a length separately.
+    /// This lets the caller pass null as the buffer, which is required when using the automatic buffer selection.
+    pub fn recv_raw(
+        self: *IO_Uring,
+        user_data: u64,
+        fd: os.fd_t,
+        buffer: ?[*]u8,
+        len: usize,
+        flags: u32,
+    ) !*io_uring_sqe {
```
```suggestion
    pub fn recv_with_automatic_buffer_selection(
        self: *IO_Uring,
        user_data: u64,
        fd: os.fd_t,
        group_id: usize,
        flags: u32,
    ) !*io_uring_sqe {
```

@vrischmann What do you think if this interface is explicitly only used for automatic buffer selection, not for normal recv/read calls where buffers are provided, since passing by slice there provides more safety in those cases?

This could then also set the `IOSQE_BUFFER_SELECT` in the prep routine, to ensure there's no confusion about where this should be done. I know this is different to how liburing does it, but I think that makes sense for C, and perhaps not Zig, which has richer types.

This interface was the one thing I wasn't sure about when I did the original io_uring PR.

kprotty replied:
Another idea for consideration: have read/recv take a `Buffer` type which is
```zig
union(enum) {
  bytes: []u8,
  iovecs: []const os.iovec,
  buffer_group_id: usize,
}
```
The variant would then call the correct prep() function. Not sure if this is any better than just having separate functions.

vrischmann replied:
Thanks for your feedback.

I like the idea of using a union for the buffer type, I was hesitant to do this because it will break all existing code using io_uring. If it's acceptable I'll get back into this PR and look at implementing it.

vrischmann commented:
I changed the PR, now `IO_Uring.read()` takes a `ReadBuffer` which can be used for simple reads or reads using buffer selection.

Both `IO_Uring.readv()` and `IO_Uring.read_fixed()` could be handled in the same way but it should be maybe go in another PR ?

jorangreef reviewed (commented):
Perhaps it makes sense to do the breaking changes in one go? Would be really nice to get `recv()` into this PR since it's so similar to `read()`.

vrischmann commented:
> Perhaps it makes sense to do the breaking changes in one go? Would be really nice to get `recv()` into this PR since it's so similar to `read()`.

Ok, I can do that.

vrischmann commented:
I've added support for buffer selection for `recv`.
I've also removed the `readv` method; `read` can now call `io_uring_prep_readv` for `ReadBuffer.iovecs`.

This means that I have to handle the `iovecs` case in `recv` and there's no API to use a `iovec` for `recv` so I simply make a slice from the first buffer in the `iovec`. 
Another option might be to just return a "not supported" error ?

kprotty commented on lib/std/os/linux/io_uring.zig:
```diff
@@ -2843,3 +2811,453 @@ test "linkat" {
+
+    const address = try net.Address.parseIp4("127.0.0.1", 3131);
+    const kernel_backlog = 1;
+    const listener_socket = try os.socket(address.any.family, os.SOCK.STREAM | os.SOCK.CLOEXEC, 0);
```
For ceremony's sake, could add `errdefer os.closeSocket(listener_socket)`. Ditto for `client` socket below.

vrischmann replied:
Thanks, done

jorangreef commented:
> I've added support for buffer selection for recv.
> I've also removed the readv method; read can now call io_uring_prep_readv for ReadBuffer.iovecs.

Thanks for doing this!

> This means that I have to handle the iovecs case in recv and there's no API to use a iovec for recv
> so I simply make a slice from the first buffer in the iovec.
> Another option might be to just return a "not supported" error ?

Rather than using `ReadBuffer` also for `recv()`, I think it would be better to use a separate, more restricted `RecvBuffer` type to eliminate this case completely.

vrischmann commented:
> Rather than using ReadBuffer also for recv(), I think it would be better to use a separate, more restricted RecvBuffer type to eliminate this case completely.
Ok makes sense, I'll change it

vrischmann commented:
I've pushed an update which uses the `RecvBuffer` type for `recv()`.

andrewrk commented:
Thanks @vrischmann and reviewers!
